include/ic_gameobject.h

```diff
@@ -11,16 +13,20 @@ namespace IC {
         GameObject(std::string name);
         ~GameObject();
 
-        std::vector<std::shared_ptr<GameObjectComponent>> &Components() { return _components; }
         std::shared_ptr<Transform> &GetTransform() { return _transform; }
-        template <typename T> void AddComponent(const T &component) {
-            _components.push_back(std::make_shared<T>(component));
+
+        template <typename T> std::shared_ptr<T> AddComponent() {
+            _componentMap[typeid(T)] = std::make_shared<T>();
```

Comment:
Maybe we'd want to make sure we're not overwriting an existing component here? Could still return the one that already exists in this case too